# Практическое задание №4

Данные data_big.csv о финских алкоголиках, переменные — признаки в первый день отказа от запоя:

Зависимая переменная - depressed.mood.1, остальные рассматриваются в качестве независимых. Независимые категориальные переменные упорядочить по значимости влияния на зависимую переменную. Привести соответствующие частоты, значимости критреия хи-квадрат и точного критерия Фишера. По совместному распределению первых двух наиболее значимых переменных вычисляете односторонний коэффициент неопределенности относительно зависимой переменной.

In [1]:
import pandas as pd
import numpy as np
import scipy
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
import warnings
warnings.filterwarnings('ignore')

Считаем данные:

In [19]:
df = pd.read_csv("data_big.csv")

In [20]:
df.head(10)

Unnamed: 0  X.1  depressed.mood.1  anxiety.1  suspiciousness.1  \
0           1    1                 1          1                 0   
1           2    2                 1          1                 0   
2           3    3                 1          1                 0   
3           4    4                 2          2                 0   
4           5    5                 1          1                 0   
5           6    6                 1          1                 0   
6           7    7                 1          1                 0   
7           8    8                 1          1                 0   
8           9    9                 1          1                 0   
9          10   10                 1          1                 0   

   irritability.1  craving.to.alcohol.1  weakness.1  insomia.1  headache.1  \
0               1                     1           1          1           0   
1               0                     1           1          2           1   
2               0                     0           2          1           0   
3               0                     0           2          0           0   
4               0                     2           2          1           0   
5               1                     1           2          0           0   
6               1                     1           2          2           1   
7               1                     0           1          2           1   
8               0                     2           1          0           0   
9               1                     1           1          2           1   

   ...  combined2.9  HR.9  SBP.9  DBP.9  MBP.9   SV.9  CO.9  SI.9  CI.9  \
0  ...            0  68.0  108.0   70.0   83.0   74.0   5.0  40.0   2.7   
1  ...            9  63.0  114.0   70.0   85.0  123.0   7.7  60.0   3.8   
2  ...            0  64.0  120.0   80.0   93.0  106.0   6.8  55.0   3.5   
3  ...            0  56.0  124.0   90.0  101.0   90.0   5.1  43.0   2.4   
4  ...            0  66.0  116.0   78.0   90.0   90.0   5.9  45.0   2.2   
5  ...            0  62.0  142.0   90.0  107.0  113.0   7.0  56.0   3.4   
6  ...            0  85.0  110.0   84.0   93.0   46.0   3.9  21.0   1.8   
7  ...            0  68.0  134.0   88.0  103.0   74.0   6.4  33.0   2.9   
8  ...            0  76.0  134.0   76.0   96.0  116.0   8.6  60.0   4.4   
9  ...            0  64.0  110.0   70.0   83.0   43.0   2.7  27.0   1.7   

    TPR.9  
0  1321.0  
1   879.0  
2  1098.0  
3  1600.0  
4  1228.0  
5  1225.0  
6  1899.0  
7  1290.0  
8   886.0  
9  2467.0  

[10 rows x 121 columns]

In [21]:
df = df[df.columns[2:20]]

In [22]:
#df = df.drop(columns=['suspiciousness.1', 'polyuria.1', 'diarrhea.1']);

In [23]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 18 columns):
 #   Column                      Non-Null Count  Dtype
---  ------                      --------------  -----
 0   depressed.mood.1            34 non-null     int64
 1   anxiety.1                   34 non-null     int64
 2   suspiciousness.1            34 non-null     int64
 3   irritability.1              34 non-null     int64
 4   craving.to.alcohol.1        34 non-null     int64
 5   weakness.1                  34 non-null     int64
 6   insomia.1                   34 non-null     int64
 7   headache.1                  34 non-null     int64
 8   tremor.1                    34 non-null     int64
 9   polyuria.1                  34 non-null     int64
 10  sweating.1                  34 non-null     int64
 11  transient.hallusinations.1  34 non-null     int64
 12  vomiting.1                  34 non-null     int64
 13  anoreksia.1                 34 non-null     int64
 14  diarrhea.1  

In [24]:
df.columns

Index(['depressed.mood.1', 'anxiety.1', 'suspiciousness.1', 'irritability.1',
       'craving.to.alcohol.1', 'weakness.1', 'insomia.1', 'headache.1',
       'tremor.1', 'polyuria.1', 'sweating.1', 'transient.hallusinations.1',
       'vomiting.1', 'anoreksia.1', 'diarrhea.1', 'thirst.1', 'chest.pain.1',
       'hyperemia.1'],
      dtype='object')

In [25]:
for col in df.columns[1:]:
    df[col] = [0 if value == 2 else value for value in df[col]]

Определим нужные функции для дальнейших вычислений:

In [26]:
def entropy(x):
    p = x/np.sum(x)
    p = p[p != 0]
    return - np.sum(p * np.log2(p))

In [27]:
def uncertian(tab):
    tab = np.array(tab)
    Hxy = entropy(tab.flatten())
    Hx = entropy(np.sum(tab, axis=1))
    Hy = entropy(np.sum(tab, axis=0))
    I = Hy + Hx - Hxy
    return I/Hx*100, I/Hy*100, 2*I/(Hx+Hy)*100

In [28]:
def chi2P(tab):
    tab = np.array(tab)
    r = np.sum(tab, axis=1)
    c = np.sum(tab, axis=0)
    chi2_ = 0

    for j in range(tab.shape[1]):
        for i in range(tab.shape[0]):
            chi2_ += (tab[i, j]**2) / (r[i] * c[j])

    chi2_ = (chi2_ - 1) * np.sum(tab)
    p_value = 1 - scipy.stats.chi2.cdf(chi2_, (tab.shape[0] - 1) * (tab.shape[1] - 1))

    return p_value

Выичслим метрики для каждой пары зависимой и независимой переменной

In [36]:
X = np.array(df["depressed.mood.1"])
print(X)
print("Энтропия признака "+"depressed.mood.1"+f": {entropy(pd.DataFrame(X).value_counts())}")
print("\n")
sig = {}
for feature in df.columns[1:]:
    print("Название признака: " + feature)
    Y = np.array(df[feature])
    print(Y)
    tab = pd.crosstab(X, Y)
    #tab = tab.T
    print(tab)
    print(f"Условыне вероятности: {tab.iloc[:,1] / tab.sum(axis=1)}")
    res = scipy.stats.fisher_exact(tab)
    sig[feature] = res[1]
    print(f"Точный критерий Фишера: {res[1]}")
    print(f"Точный критерий Пирсона: {chi2P(tab)}")
    print("Энтропия признака "+feature+f": {entropy(pd.DataFrame(Y).value_counts())}")
    print(f"Коэффициенты неопределенности: {uncertian(tab)}")
    print("\n")

[1 1 1 2 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 1 1 1 1 1 2 1 2 1 2 1 1 1 2]
Энтропия признака depressed.mood.1: 0.6722948170756379


Название признака: anxiety.1
[1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 0 0 0 0 0 1 0 1 0]
col_0  0   1
row_0       
1      4  24
2      5   1
Условыне вероятности: row_0
1    0.857143
2    0.166667
dtype: float64
Точный критерий Фишера: 0.00240463259831185
Точный критерий Пирсона: 0.0005033319307589101
Энтропия признака anxiety.1: 0.833764907210665
Коэффициенты неопределенности: (34.478197199470316, 27.801018103366836, 30.781665435346607)


Название признака: suspiciousness.1
[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1]
col_0   0  1
row_0       
1      28  0
2       5  1
Условыне вероятности: row_0
1    0.000000
2    0.166667
dtype: float64
Точный критерий Фишера: 0.1764705882352941
Точный критерий Пирсона: 0.028326575664195386
Энтропия признака suspiciousness.1: 0.19143325481419343
Коэффициенты неопределенности: (11.4121682558

### Какие переменные наиболее значимые?

In [37]:
sorted(sig.items(), key=lambda x: x[1])

[('anxiety.1', 0.00240463259831185),
 ('hyperemia.1', 0.008244454622783486),
 ('weakness.1', 0.023895385841666027),
 ('insomia.1', 0.1478276516390761),
 ('tremor.1', 0.17336330325435867),
 ('thirst.1', 0.17484073212660534),
 ('suspiciousness.1', 0.1764705882352941),
 ('transient.hallusinations.1', 0.32620320855614976),
 ('diarrhea.1', 0.32620320855614976),
 ('chest.pain.1', 0.37005912689678966),
 ('sweating.1', 0.38719789665284665),
 ('anoreksia.1', 0.38719789665284665),
 ('irritability.1', 0.3891088137145848),
 ('vomiting.1', 0.5798138751910916),
 ('headache.1', 0.6378737813256559),
 ('craving.to.alcohol.1', 0.6561836383860856),
 ('polyuria.1', 1.0)]

При отсутствии депрессии тревожность наблюдается в 17% случаев, а при её наличии более чем в 85%. Различие в частотах нельзя объяснить случайностью, p_value = 0.0024. Для признака "тревожность" неопределенность меньше, чем для признака "депрессия". Есди депрессия не сильная, тревожности скорее всего нет.